File: Algorithms Analysis/Labs/Lab 3 G1-4/HashingLab2B.py
import time
import math

class HashTable:
    def __init__(self, size, mode="LP"):
        self.capacity = size
        self.mode = mode # "LP" for Linear Probing, "DH" for Double Hashing
        self.table = [None] * self.capacity # Elements stored as (key, "OCCUPIED") or (key, "DELETED")
        self.KNUTH_A = 0.6180339887

    def _hash1(self, key):
        # TODO: Implement primary hash function using the Multiplication Method
        return 0

    def _hash2(self, key):
        # TODO: Implement secondary hash function for Double Hashing
        # Must always produce an odd number
        return 1

    def insert(self, key):
        # TODO: Implement insertion logic handling OCCUPIED and DELETED spots
        pass

    def search(self, key):
        probes = 0
        found = False
        
        # TODO: Implement successful search logic.
        # CRITICAL: Increment 'probes' every single time you look at a slot index.
        # Ensure you scan right past "DELETED" tokens.
        
        return found, probes

    def remove(self, key):
        # TODO: Implement lazy deletion by tagging the slot status as "DELETED"
        pass


def generate_keys(count):
    return [i * 7 + 3 for i in range(1, count + 1)]


def main():
    m = 1024 # Assigned power-of-2 table size
    
    with open("data.txt", "w") as f:
        f.write("#Alpha\tTime_LP\tTime_DH\tProbes_LP\tProbes_DH\n")
        
        print("Starting empirical profiling for Flavor 2 (Lazy Deletion Overhead)...")
        
        # Stepping active load factor alpha from 0.1 to 0.45
        for step in range(2, 10):
            alpha = (step * 5) / 100.0
            active_elements = int(m * alpha)
            total_to_insert = active_elements * 2
            
            keys = generate_keys(total_to_insert)
            
            # ------------------------------------------
            # PROFILE LINEAR PROBING
            # ------------------------------------------
            lp_table = HashTable(m, mode="LP")
            for key in keys:
                lp_table.insert(key)
            for i in range(0, total_to_insert, 2):
                lp_table.remove(keys[i]) # Delete half
                
            lp_probes = 0
            start_lp = time.perf_counter()
            for i in range(1, total_to_insert, 2): # Search remaining keys
                _, p = lp_table.search(keys[i])
                lp_probes += p
            stop_lp = time.perf_counter()
            duration_lp = (stop_lp - start_lp) * 1_000_000
            avg_probes_lp = lp_probes / active_elements
            
            # ------------------------------------------
            # PROFILE DOUBLE HASHING
            # ------------------------------------------
            dh_table = HashTable(m, mode="DH")
            for key in keys:
                dh_table.insert(key)
            for i in range(0, total_to_insert, 2):
                dh_table.remove(keys[i])
                
            dh_probes = 0
            start_dh = time.perf_counter()
            for i in range(1, total_to_insert, 2):
                _, p = dh_table.search(keys[i])
                dh_probes += p
            stop_dh = time.perf_counter()
            duration_dh = (stop_dh - start_dh) * 1_000_000
            avg_probes_dh = dh_probes / active_elements
            
            f.write(f"{alpha:.2f}\t{duration_lp:.2f}\t{duration_dh:.2f}\t{avg_probes_lp:.2f}\t{avg_probes_dh:.2f}\n")
            print(f"Finished processing Active Load Factor (\u03B1): {alpha:.2f}")

    print("Profiling completed successfully. Results recorded in data.txt")

if __name__ == "__main__":
    main()

File: Algorithms Analysis/Labs/Lab 3 G1-4/HashingLab2B.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <fstream>
#include <cmath>

using namespace std;
using namespace std::chrono;

enum Mode { LINEAR_PROBING, DOUBLE_HASHING };
enum SlotState { EMPTY, OCCUPIED, DELETED };

struct HashEntry {
    int key;
    SlotState state = EMPTY;
};

class HashTable {
private:
    int capacity;
    Mode mode;
    vector<HashEntry> table;
    const double KNUTH_A = 0.6180339887; // Constant A for multiplication method

    int hash1(int key) {
        // TODO: Implement primary hash function using the Multiplication Method
        // Formula: floor(m * ((key * A) mod 1))
        return 0; 
    }

    int hash2(int key) {
        // TODO: Implement secondary hash function for Double Hashing
        // Must return an odd number to be relatively prime to power-of-2 table size
        return 1; 
    }

public:
    HashTable(int size, Mode m) : capacity(size), mode(m) {
        table.resize(capacity);
    }

    // Inserts a key into the table. 
    void insert(int key) {
        // TODO: Implement insertion logic.
        // Remember to reuse DELETED slots appropriately depending on your search rules.
    }

    // Searches for a key. Returns a pair: <found (bool), probes_counted (int)>
    pair<bool, int> search(int key) {
        int probes = 0;
        bool found = false;
        
        // TODO: Implement search logic.
        // CRITICAL: Increment 'probes' every single time you check a slot.
        // Note: You must scan past DELETED slots during a search!
        
        return {found, probes};
    }

    // Deletes a key by replacing its state with DELETED
    void remove(int key) {
        // TODO: Implement lazy deletion logic
    }
};

// Generates uniform random or sequential data keys
vector<int> generateKeys(int count) {
    vector<int> data;
    for (int i = 1; i <= count; i++) {
        data.push_back(i * 7 + 3); // Basic distinct keys
    }
    return data;
}

int main() {
    int m = 1024; // Assigned power-of-2 table size
    
    ofstream outFile("data.txt");
    outFile << "#Alpha\tTime_LP\tTime_DH\tProbes_LP\tProbes_DH\n";

    cout << "Starting empirical profiling for Flavor 2 (Lazy Deletion Overhead)..." << endl;

    // Loop through target active load factors (0.1 to 0.45)
    // We stop early because we insert 2x elements and delete half to create the sentinels
    for (double alpha = 0.1; alpha <= 0.45; alpha += 0.05) {
        int activeElements = static_cast<int>(m * alpha);
        int totalToInsert = activeElements * 2; // Insert double the elements
        
        vector<int> keys = generateKeys(totalToInsert);

        // ==========================================
        // PROFILE LINEAR PROBING
        // ==========================================
        HashTable lpTable(m, LINEAR_PROBING);
        for (int key : keys) {
            lpTable.insert(key);
        }
        // Delete half the keys to pollute the paths with DELETED sentinels
        for (int i = 0; i < totalToInsert; i += 2) {
            lpTable.remove(keys[i]);
        }
        
        // Run Successful Searches on the remaining keys
        long long lpTotalProbes = 0;
        auto startLP = high_resolution_clock::now();
        for (int i = 1; i < totalToInsert; i += 2) {
            auto result = lpTable.search(keys[i]);
            lpTotalProbes += result.second;
        }
        auto stopLP = high_resolution_clock::now();
        auto durationLP = duration_cast<microseconds>(stopLP - startLP).count();
        double avgProbesLP = (double)lpTotalProbes / activeElements;

        // ==========================================
        // PROFILE DOUBLE HASHING
        // ==========================================
        HashTable dhTable(m, DOUBLE_HASHING);
        for (int key : keys) {
            dhTable.insert(key);
        }
        for (int i = 0; i < totalToInsert; i += 2) {
            dhTable.remove(keys[i]);
        }
        
        long long dhTotalProbes = 0;
        auto startDH = high_resolution_clock::now();
        for (int i = 1; i < totalToInsert; i += 2) {
            auto result = dhTable.search(keys[i]);
            dhTotalProbes += result.second;
        }
        auto stopDH = high_resolution_clock::now();
        auto durationDH = duration_cast<microseconds>(stopDH - startDH).count();
        double avgProbesDH = (double)dhTotalProbes / activeElements;

        // Export data row
        outFile << alpha << "\t\t" << durationLP << "\t\t" << durationDH << "\t\t" 
                << avgProbesLP << "\t\t\t" << avgProbesDH << "\n";
                
        cout << "Finished processing Active Load Factor (\u03B1): " << alpha << endl;
    }

    outFile.close();
    cout << "Profiling completed successfully. Results recorded in data.txt" << endl;
    return 0;
}